drape/glyph_manager.cpp

```diff
@@ -659,46 +626,68 @@ hb_language_t OrganicMapsLanguageToHarfbuzzLanguage(int8_t lang)
 }
 }  // namespace
 
+// This method is NOT multithreading-safe.
 text::TextMetrics GlyphManager::ShapeText(std::string_view utf8, int fontPixelHeight, int8_t lang)
 {
+#ifdef DEBUG
+  static int const fontSize = fontPixelHeight;
+  ASSERT_EQUAL(fontSize, fontPixelHeight,
+      ("Cache relies on the same font height/metrics for each glyph", fontSize, fontPixelHeight));
+#endif
+
+  // A simple cache greatly speeds up text metrics calculation. It has 80+% hit ratio in most scenarios.
+  if (auto const found = m_impl->m_textMetricsCache.find(utf8); found != m_impl->m_textMetricsCache.end())
+    return found->second;
+
   const auto [text, segments] = harfbuzz_shaping::GetTextSegments(utf8);
 
   // TODO(AB): Optimize language conversion.
   hb_language_t const hbLanguage = OrganicMapsLanguageToHarfbuzzLanguage(lang);
 
   text::TextMetrics allGlyphs;
 
-  // TODO(AB): Cache runs.
-  // TODO(AB): Cache buffer.
-  hb_buffer_t * buf = hb_buffer_create();
   for (auto const & substring : segments)
   {
-    hb_buffer_clear_contents(buf);
+    hb_buffer_clear_contents(m_impl->m_harfbuzzBuffer);
 
     // TODO(AB): Some substrings use different fonts.
-    hb_buffer_add_utf16(buf, reinterpret_cast<const uint16_t *>(text.data()), static_cast<int>(text.size()),
-                        substring.m_start, substring.m_length);
-    hb_buffer_set_direction(buf, substring.m_direction);
-    hb_buffer_set_script(buf, substring.m_script);
-    hb_buffer_set_language(buf, hbLanguage);
-
-    // TODO(AB): Check not only the first character to determine font for the run, but all chars.
-    auto firstCharacterIter{text.begin() + substring.m_start};
-    auto const firstCharacterUnicode = utf8::unchecked::next16(firstCharacterIter);
-
-    int const fontIndex = GetFontIndex(firstCharacterUnicode);
-    if (fontIndex < 0)
+    hb_buffer_add_utf16(m_impl->m_harfbuzzBuffer, reinterpret_cast<const uint16_t *>(text.data()),
+        static_cast<int>(text.size()), substring.m_start, substring.m_length);
+    hb_buffer_set_direction(m_impl->m_harfbuzzBuffer, substring.m_direction);
+    hb_buffer_set_script(m_impl->m_harfbuzzBuffer, substring.m_script);
+    hb_buffer_set_language(m_impl->m_harfbuzzBuffer, hbLanguage);
+
+    auto u32CharacterIter{text.begin() + substring.m_start};
+    auto const end{text.begin() + substring.m_start + substring.m_length};
```

Comment:
Do you really like this {} notation instead of a natural language wise assignment?
```auto const end = u32CharacterIter + substring.m_length;```

Reply:
This notation helps to discover any type-conversion issues.

Reply:
Isn't it the same equivalent value initialization here? What warnings?
```
auto a{b};
auto a = b;
```

Reply:
Here it's ok, no warnings. I can change it to = if you insist )

Braces initialization has many advantages and a few disadvantages in some specific corner cases: https://stackoverflow.com/questions/18222926/what-are-the-advantages-of-list-initialization-using-curly-braces